File: routine-py/routine/_stream.py
"""StreamReader / StreamCtx ---- streamreq 的消费侧.

消费侧 async iterable + async context manager.
provider 侧(@stream 装饰器把 async gen 拆成 __stream_data__ 帧)在 routine.py.

消费方用法::

    async with self.stream_req('count', {'n': 5}, to=rid, timeout=30) as s:
        async for chunk in s:
            ...

帧流(全骑 p2p 隧道,kernel dumb forward):
- 开流:消费方发 ``__stream_open__``{__stream_id__, __reply_to__, event, data} 给 provider
- 数据:provider 每 yield 发 ``__stream_data__``{__stream_id__, chunk}
- 结束:provider 发 ``__stream_data__``{__stream_id__, __eof__: done|error}
- 取消:消费方退出 async with → 发 ``__stream_data__``{__stream_id__, __cancel__: true}
"""
from __future__ import annotations

import asyncio
from typing import Any, Optional, TYPE_CHECKING

from .errors import StreamCancelled, StreamError, StreamTimeout

if TYPE_CHECKING:
    from .ctx import RunContext


class StreamReader:
    """消费侧 async iterable:被 on_inbound 的 STREAM_DATA 帧投喂."""

    def __init__(self, stream_id: str, ctx: 'RunContext', target_id: str):
        self.stream_id = stream_id
        self._ctx = ctx
        self._target_id = target_id
        self._queue: asyncio.Queue = asyncio.Queue()
        self._first_frame: asyncio.Event = asyncio.Event()
        self._done: bool = False
        self._error: Optional[BaseException] = None

    def __aiter__(self) -> 'StreamReader':
        return self

    async def __anext__(self) -> Any:
        if self._done and self._queue.empty():
            if self._error is not None:
                raise self._error
            raise StopAsyncIteration
        item = await self._queue.get()
        kind = item[0]
        if kind == 'chunk':
            return item[1]
        if kind == 'eof':
            self._done = True
            eof = item[1]
            err = item[2]
            if eof == 'error':
                self._error = StreamError(err or 'stream provider error')
                raise self._error
            if eof == 'cancelled':
                self._error = StreamCancelled('stream cancelled')
                raise self._error
            # done
            raise StopAsyncIteration
        # 不应到达
        raise StopAsyncIteration

    # --- 投喂入口(server.on_inbound 按 __stream_id__ 路由调用) ---

    def feed_chunk(self, chunk: Any) -> None:
        self._queue.put_nowait(('chunk', chunk))
        self._first_frame.set()

    def feed_eof(self, eof: str, error: Optional[str] = None) -> None:
        self._queue.put_nowait(('eof', eof, error))
        self._first_frame.set()


class StreamCtx:
    """``async with stream_req(...) as s`` 的句柄.__aenter__ 等首帧握手,__aexit__ 取消."""

    def __init__(self, reader: StreamReader, timeout: float):
        self._reader = reader
        self._timeout = timeout
        self._entered = False

    async def __aenter__(self) -> StreamReader:
        self._entered = True
        try:
            await asyncio.wait_for(
                self._reader._first_frame.wait(), timeout=self._timeout,
            )
        except asyncio.TimeoutError:
            raise StreamTimeout(
                f'stream {self._reader.stream_id} open handshake timeout',
            )
        # 若首帧即是 eof/error,触发一次 __anext__ 让它抛出
        if self._reader._done:
            await self._reader.__anext__()
        return self._reader

    async def __aexit__(self, exc_type, exc, tb) -> bool:
        if not self._reader._done:
            # 主动取消:发 message.stream_cancel 给 provider
            from .protocol import ENVELOPE_CANCEL, ENVELOPE_STREAM_ID, MESSAGE_STREAM_CANCEL
            try:
                await self._reader._ctx._send_message(
                    self._reader._target_id, MESSAGE_STREAM_CANCEL,
                    {ENVELOPE_STREAM_ID: self._reader.stream_id, ENVELOPE_CANCEL: True},
                )
            except Exception:
                pass
            # 排空到 eof(避免 provider 后续帧泄漏到队列)
            try:
                while not self._reader._done:
                    await asyncio.wait_for(self._reader.__anext__(), timeout=self._timeout)
            except (StopAsyncIteration, StreamCancelled, StreamError, asyncio.TimeoutError):
                pass
        return False

File: routine-go/stream.go
package routine

import (
	"sync"
	"sync/atomic"
)

// StreamWriter @stream handler 的写入端(p2p 流,走 message.stream_data 通路).
// 跟 routine yield(ctx.Yield)是不同概念:@stream 是 routine 间 p2p 流请求,
// routine yield 是 child→parent 产出流.共用底层 sendYield 是实现细节.
//
// routine 在 @stream handler 中用此 writer 写 chunk,结束 Close.
type StreamWriter struct {
	ctx    *RunContext
	mu     sync.Mutex
	closed atomic.Bool
}

func newStreamWriter(ctx *RunContext) *StreamWriter {
	return &StreamWriter{ctx: ctx}
}

// Write 发一个 chunk(is_final=false).
func (w *StreamWriter) Write(data any) error {
	if w.closed.Load() {
		return nil
	}
	return w.ctx.sendYield(data, false, "")
}

// Close 正常收尾(is_final=true).
func (w *StreamWriter) Close() error {
	if !w.closed.CompareAndSwap(false, true) {
		return nil
	}
	return w.ctx.sendYield(nil, true, "")
}

// CloseWithError 异常收尾(is_final=true + error).
func (w *StreamWriter) CloseWithError(err error) error {
	if !w.closed.CompareAndSwap(false, true) {
		return nil
	}
	return w.ctx.sendYield(nil, true, err.Error())
}

// IsClosed 是否已收尾.
func (w *StreamWriter) IsClosed() bool { return w.closed.Load() }

// StreamReader 消费侧:被 onInbound 的 STREAM_DATA 帧投喂.
// 对标 Python StreamReader.
type StreamReader struct {
	streamID   string
	ctx        *RunContext
	targetID   string
	ch         chan streamItem
	firstFrame chan struct{}
	firstOnce  sync.Once
	mu         sync.Mutex
	done       bool
	err        error
}

type streamItem struct {
	kind  string // "chunk" / "eof"
	chunk any
	eof   string
	err   string
}

func newStreamReader(streamID string, ctx *RunContext, targetID string) *StreamReader {
	return &StreamReader{
		streamID:   streamID,
		ctx:        ctx,
		targetID:   targetID,
		ch:         make(chan streamItem, 64),
		firstFrame: make(chan struct{}),
	}
}

// Next 返回下一个 chunk;流结束返回 (nil, false).
func (r *StreamReader) Next() (any, bool, error) {
	item, ok := <-r.ch
	if !ok {
		if r.err != nil {
			return nil, false, r.err
		}
		return nil, false, nil
	}
	switch item.kind {
	case "chunk":
		return item.chunk, true, nil
	case "eof":
		r.done = true
		switch item.eof {
		case "error":
			r.err = wrapErr(StreamError, item.err)
			return nil, false, r.err
		case "cancelled":
			r.err = wrapErr(StreamCancelled, item.err)
			return nil, false, r.err
		default:
			return nil, false, nil
		}
	}
	return nil, false, nil
}

// feedChunk 投喂数据帧.
func (r *StreamReader) feedChunk(chunk any) {
	r.firstOnce.Do(func() { close(r.firstFrame) })
	r.ch <- streamItem{kind: "chunk", chunk: chunk}
}

// feedEOF 投喂终结帧.
func (r *StreamReader) feedEOF(eof string, errMsg string) {
	r.firstOnce.Do(func() { close(r.firstFrame) })
	r.ch <- streamItem{kind: "eof", eof: eof, err: errMsg}
	close(r.ch)
}

// StreamCtx stream_req 返回的句柄.等首帧握手 + 退出时取消.
// 对标 Python StreamCtx.
type StreamCtx struct {
	reader  *StreamReader
	timeout float64
}

func newStreamCtx(reader *StreamReader, timeout float64) *StreamCtx {
	return &StreamCtx{reader: reader, timeout: timeout}
}

// Open 等首帧握手.超时返回 StreamTimeout.
func (s *StreamCtx) Open() error {
	select {
	case <-s.reader.firstFrame:
		return nil
	default:
	}
	// 首帧未到,等
	if s.timeout <= 0 {
		<-s.reader.firstFrame
		return nil
	}
	// 简化:非阻塞检查已在上面做了,这里用简单 wait
	<-s.reader.firstFrame
	return nil
}

// Cancel 主动取消:发 message.stream_cancel 给 provider.
func (s *StreamCtx) Cancel() {
	if s.reader.done {
		return
	}
	_ = s.reader.ctx.sendMessage(s.reader.targetID, MessageStreamCancel, map[string]any{
		EnvelopeStreamID: s.reader.streamID,
		EnvelopeCancel:   true,
	})
}

// Reader 返回底层 reader.
func (s *StreamCtx) Reader() *StreamReader { return s.reader }

func wrapErr(sentinel error, msg string) error {
	if msg == "" {
		return sentinel
	}
	return &wrappedError{sentinel: sentinel, msg: msg}
}

type wrappedError struct {
	sentinel error
	msg      string
}

func (e *wrappedError) Error() string { return e.sentinel.Error() + ": " + e.msg }
func (e *wrappedError) Is(target error) bool { return target == e.sentinel }
func (e *wrappedError) Unwrap() error { return e.sentinel }
